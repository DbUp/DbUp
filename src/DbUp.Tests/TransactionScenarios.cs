using DbUp.Builder;
using DbUp.Engine;
using DbUp.Engine.Output;
using DbUp.Tests.TestInfrastructure;
using NUnit.Framework;
using Shouldly;
using TestStack.BDDfy;

namespace DbUp.Tests
{
    [TestFixture]
    public class TransactionScenarios
    {
        UpgradeEngineBuilder upgradeEngineBuilder;
        RecordingDbConnection testConnection;
        SqlScript[] scripts;
        CaptureLogsLogger logger;

        [SetUp]
        public void CaptureLogs()
        {
            logger = new CaptureLogsLogger();
        }

        [Test]
        public void UsingNoTransactionsScenario()
        {
            this
                .Given(_ => DbUpSetupToNotUseTransactions())
                .When(_ => UpgradeIsPerformedExecutingTwoScripts())
                .Then(_ => ShouldExecuteScriptsWithoutUsingATransaction())
                .BDDfy();
        }

        [Test]
        public void UsingNoTransactionsScenarioScriptFails()
        {
            this
                .Given(_ => DbUpSetupToNotUseTransactions())
                .When(_ => UpgradeIsPerformedWithFirstOfTwoScriptsFails())
                .Then(_ => ShouldStopExecution())
                .BDDfy();
        }

        [Test]
        public void UsingTransactionPerScriptScenarioSuccess()
        {
            this
                .Given(_ => DbUpSetupToUseTransactionPerScript())
                .When(_ => UpgradeIsPerformedExecutingTwoScripts())
                .Then(_ => ShouldHaveExecutedEachScriptInATransaction())
                .BDDfy();
        }

        [Test]
        public void UsingTransactionPerScriptScenarioScriptFails()
        {
            this
                .Given(_ => DbUpSetupToUseTransactionPerScript())
                .When(_ => UpgradeIsPerformedWithFirstOfTwoScriptsFails())
                .Then(_ => ShouldRollbackFailedScriptAndStopExecution())
                .BDDfy();
        }

        [Test]
        public void UsingSingleTransactionScenarioSuccess()
        {
            this
                .Given(_ => DbUpSetupToUseSingleTransaction())
                .When(_ => UpgradeIsPerformedExecutingTwoScripts())
                .Then(_ => ShouldExecuteAllScriptsInASingleTransaction())
                .BDDfy();
        }

        [Test]
        public void UsingSingleTransactionScenarioSuccessScriptFails()
        {
            this
                .Given(_ => DbUpSetupToUseSingleTransaction())
                .When(_ => UpgradeIsPerformedWithFirstOfTwoScriptsFails())
                .Then(_ => ShouldRollbackFailedScriptAndStopExecution())
                .BDDfy();
        }

        void UpgradeIsPerformedWithFirstOfTwoScriptsFails()
        {
            scripts = new[]
            {
                new SqlScript("Script0001.sql", "error"),
                new SqlScript("Script0002.sql", "print 'script2'")
            };
            upgradeEngineBuilder
                .WithScripts(scripts)
                .Build()
                .PerformUpgrade();
        }

        void ShouldStopExecution()
        {
            logger.Log.ShouldMatchApproved(b =>
            {
                b.WithScrubber(Scrubbers.ScrubDates);
                b.LocateTestMethodUsingAttribute<TestAttribute>();
            });
        }

        void ShouldRollbackFailedScriptAndStopExecution()
        {
            logger.Log.ShouldMatchApproved(b =>
            {
                b.WithScrubber(Scrubbers.ScrubDates);
                b.LocateTestMethodUsingAttribute<TestAttribute>();
            });
        }

        void ShouldExecuteAllScriptsInASingleTransaction()
        {
            logger.Log.ShouldMatchApproved(b =>
            {
                b.WithScrubber(Scrubbers.ScrubDates);
                b.LocateTestMethodUsingAttribute<TestAttribute>();
            });
        }

        void ShouldHaveExecutedEachScriptInATransaction()
        {
            logger.Log.ShouldMatchApproved(b =>
            {
                b.WithScrubber(Scrubbers.ScrubDates);
                b.LocateTestMethodUsingAttribute<TestAttribute>();
            });
        }

        void ShouldExecuteScriptsWithoutUsingATransaction()
        {
            logger.Log.ShouldMatchApproved(b =>
            {
                b.WithScrubber(Scrubbers.ScrubDates);
                b.LocateTestMethodUsingAttribute<TestAttribute>();
            });
        }

        void DbUpSetupToUseSingleTransaction()
        {
            testConnection = new RecordingDbConnection(logger, false, "SchemaVersions");
            upgradeEngineBuilder = DeployChanges.To
                .TestDatabase(testConnection)
                .JournalToSqlTable("dbo", "SchemaVersions")
                .WithTransaction();
        }

        void DbUpSetupToNotUseTransactions()
        {
            testConnection = new RecordingDbConnection(logger, false, "SchemaVersions");
            upgradeEngineBuilder = DeployChanges.To
                .TestDatabase(testConnection)
                .JournalToSqlTable("dbo", "SchemaVersions")
                .WithoutTransaction();
        }

        void DbUpSetupToUseTransactionPerScript()
        {
            testConnection = new RecordingDbConnection(logger, false, "SchemaVersions");
            upgradeEngineBuilder = DeployChanges.To
                .TestDatabase(testConnection)
                .JournalToSqlTable("dbo", "SchemaVersions")
                .WithTransactionPerScript();
        }

        void UpgradeIsPerformedExecutingTwoScripts()
        {
            scripts = new[]
            {
                new SqlScript("Script0001.sql", "print 'script1'"),
                new SqlScript("Script0002.sql", "print 'script2'")
            };
             var result = upgradeEngineBuilder
                .WithScripts(scripts)
                .LogTo(logger)
                .Build()
                .PerformUpgrade();
        }
    }
}