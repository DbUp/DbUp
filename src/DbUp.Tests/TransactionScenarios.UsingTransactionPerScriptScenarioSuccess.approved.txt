DB Operation: Open connection
Info:         Beginning database upgrade
Info:         Checking whether journal table exists..
DB Operation: Begin transaction
DB Operation: Execute scalar command: select 1 from INFORMATION_SCHEMA.TABLES where TABLE_NAME = 'SchemaVersions' and TABLE_SCHEMA = 'dbo'
DB Operation: Dispose command
DB Operation: Commit transaction
DB Operation: Dispose transaction
Info:         Creating the [dbo].[SchemaVersions] table
DB Operation: Begin transaction
DB Operation: Execute non query command: create table [dbo].[SchemaVersions] (
    [Id] int identity(1,1) not null constraint [PK_SchemaVersions_Id] primary key,
    [ScriptName] nvarchar(255) not null,
    [Applied] datetime not null
)
DB Operation: Dispose command
Info:         The [dbo].[SchemaVersions] table has been created
DB Operation: Commit transaction
DB Operation: Dispose transaction
Info:         Fetching list of already executed scripts.
DB Operation: Begin transaction
DB Operation: Execute reader command: select [ScriptName] from [dbo].[SchemaVersions] order by [ScriptName]
DB Operation: Dispose command
DB Operation: Commit transaction
DB Operation: Dispose transaction
Info:         Executing Database Server script 'Script0001.sql'
DB Operation: Begin transaction
DB Operation: Execute non query command: print 'script1'
DB Operation: Dispose command
DB Operation: Create parameter
Info:         DB Operation: Add parameter to command: scriptName=Script0001.sql
DB Operation: Create parameter
Info:         DB Operation: Add parameter to command: applied=<date>
DB Operation: Execute non query command: insert into [dbo].[SchemaVersions] (ScriptName, Applied) values (@scriptName, @applied)
DB Operation: Dispose command
DB Operation: Commit transaction
DB Operation: Dispose transaction
Info:         Executing Database Server script 'Script0002.sql'
DB Operation: Begin transaction
DB Operation: Execute non query command: print 'script2'
DB Operation: Dispose command
DB Operation: Create parameter
Info:         DB Operation: Add parameter to command: scriptName=Script0002.sql
DB Operation: Create parameter
Info:         DB Operation: Add parameter to command: applied=<date>
DB Operation: Execute non query command: insert into [dbo].[SchemaVersions] (ScriptName, Applied) values (@scriptName, @applied)
DB Operation: Dispose command
DB Operation: Commit transaction
DB Operation: Dispose transaction
Info:         Upgrade successful
DB Operation: Dispose connection
